RIOT-OS/RIOT
# sock_async: add optional callback argument

### Contribution description
In https://github.com/RIOT-OS/RIOT/pull/12907/files#r390451905 I noticed that there is simply no way to get the surrounding DTLS sock from within the UDP sock's asynchronous callback. As such provides this change a way to provide context argument with the callback, similar as we already do it for `xtimer` or other asynchronous operations (and e.g. `lwip` does to for the asynchronous component of its `sock` equivalent `netconn`).

As `sock_async` is still experimental this API change should only require one ACK.




### Testing procedure
The following applications should still compile for all supported boards

- `examples/gcoap`
- `tests/gnrc_sock_async_event`
- `tests/lwip`

I amended the tests for GNRC's and lwIP's `sock_async` implementations if the callback argument supplied is correctly handed to the callback. As such the following tests should still pass

- `tests/gnrc_sock_async_event`
- `tests/lwip`

(`make -C tests/lwip tests` only succeeds on `native` and requires to have `dist/tools/tapsetup/tapsetup` run first)




### Issues/PRs references
Requirement for this PR came up in https://github.com/RIOT-OS/RIOT/pull/12907/files#r390451905. As such, #12907 depends on this PR.
